Update: false positives in function-call-argument-newline (fixes #12123)

scottohara opened the pull request:
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[X] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**What changes did you make? (Give an overview)**

Fixes #12123

Given two adjacent arguments (`a` & `b`), the rule logic was previously:
1. let `prevArgToken` be the _first_ token of `a`
2. let `currentArgToken` be the _first_ token of `b`
3. compare the _start_ line number of `prevArgToken` with the _start_ line number of `currentArgToken`

This meant that the following examples were treated as invalid, despite the linebreaks **between** arguments being correct for the specified option:

```js
/*eslint function-call-argument-newline:["error", "never"] */
fn({
 a: 1
}, b);

/*eslint function-call-argument-newline:["error", "consistent"] */
fn({
 a: 1
}, b, c);
```

Changed the rule logic to be:
1. let `prevArgToken` be the _last_ token of `a`
2. let `currentArgToken` be the _first_ token of `b`
3. compare the _end_ line number of `prevArgToken` with the _start_ line number of `currentArgToken`

Valid & invalid tests added for each of the three options (`always`, `never`, `consistent`).

All new & all existing unit tests are passing.

**Is there anything you'd like reviewers to focus on?**

Nothing in particular

mdjermanovic commented:
Hi @scottohara, thanks for the PR!

Just to note that #12123 isn't accepted as a bug yet. It could be the correct behavior as well, or the rule might need an additional option to further customize the behavior in these cases with multiline arguments.

The proposed solution seems logical and I believe that it solves the problem from #12123. But, it can produce **more** warnings in some other cases, so I'm labeling this as an enhancement for now since it has to be evaluated.

klimashkin commented:
@mdjermanovic What is the evaluation process? #12123 has been closed by the bot

mdjermanovic commented:
It's reopened now

mdjermanovic commented:
I forgot this - the PR title should also contain the name of this rule

scottohara commented:
Thanks for the review @mdjermanovic.

I've added additional tests for the multi-line template string case as requested, and updated the PR title

With a long rule name like `function-call-argument-newline`, it was hard to get the title under 72-chars while still including both the `Update:` prefix and a `(fixes #xxx)` issue reference.  Hopefully the current title is descriptive enough.
